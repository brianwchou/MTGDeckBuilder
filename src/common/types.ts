export interface CardInfo {
    artist: string,
    cmc: number,
    color_identity: Array<string>,
    colors: Array<string>,
    image_uris: {
        art_crop: string,
        border_crop: string,
        large: string,
        normal: string,
        png: string,
        small: string
    },
    mana_cost: string,
    name: string,
    oracle_text: string,
    power: string,
    rarity: string,
    reserved: boolean,
    setName: string,
    toughness: string,
    typeLine: string,
}