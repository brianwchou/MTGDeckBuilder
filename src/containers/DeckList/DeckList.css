.right-align {
    text-align: right;
}
.center-align {
    text-align: left;
}

table, th, td {
    border: 1px solid black;
}

table {
    width: 100%
}