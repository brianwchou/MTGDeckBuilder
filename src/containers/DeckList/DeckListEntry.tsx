import React from 'react';

export default function DeckListEntry ({index, card, count, getCardInfo}) {
    const handleOnClick = (e) => {
        getCardInfo(card, e.target.name);
    }

    return (
        <div className="flex alignment" key={index}> 
            <div className="flex count">
                {count}x 
                <button onClick={handleOnClick} name={'increment'}>+</button>
            </div>
            <div className="flex name">
                {card.name} 
                <button onClick={handleOnClick} name={'decrement'}>-</button>
                <button onClick={handleOnClick} name={'maybe'}>maybe</button>
            </div>        
        </div>
    )
}