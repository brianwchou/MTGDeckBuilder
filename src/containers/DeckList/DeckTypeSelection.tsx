import React from 'react'

export default function DeckTypeSelection() {
    return (
        <select>
            <option value="standard">Standard</option>
            <option value="modern">Modern</option>
            <option value="legacy">Legacy</option>
            <option value="vintage">Vintage</option>
            <option value="commander">Commander</option>
            <option value="other">Other</option>
        </select>
    )
} 
