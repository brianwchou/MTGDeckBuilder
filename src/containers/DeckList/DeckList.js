import React from 'react';
import { connect } from 'react-redux';
import './DeckList.css'

const mapStateToProps = (state) => {
    return {
        main: state.deckList.main,
        counts: state.deckList.cardCount
    }
}

const decklistStyle = {
    border: 'solid black thin',
    width: '35vw',
    height: '50vh',
    overflow: 'scroll',
}

const DeckListEntry = ({key, name, count}) => {
    return (
        <tr key={key}> 
            <td>{name}</td>
            <td className={'center-align'}>{count}</td>  
            <td className={'right-align'}>
                <button>plus</button>
                /
                <button>minus</button>
            </td>
        </tr>
    )
}

// presentational
const DeckListEntries = ({data, type, counts}) => {

    let entries = data.map((entry, index)=> {
        return <DeckListEntry key={index} name={entry.name} count={counts[entry.name]} />
    }) 
    return (entries.length) ? (
        <>
            <span><b>{type}</b></span>
            {entries}
        </>
    ) : null;
}

// presentational
const DeckTypeSelection = () => {
    return (
        <select>
            <option hidden disabled selected value> -- select a format -- </option>
            <option value="standard">Standard</option>
            <option value="modern">Modern</option>
            <option value="legacy">Legacy</option>
            <option value="vintage">Vinatage</option>
            <option value="commander">Commander</option>
            <option value="other">Other</option>
        </select>
    )
} 

//conatiner
class DeckList extends React.Component {
    render() {
        // expecting deck lists sort data here?
        const sortedByTypes = this.props.main.reduce((sortedByTypes, cardData) => {
            if (cardData.typeLine.toLowerCase().includes('creature')) {
                return {...sortedByTypes, creatures: [...sortedByTypes.creatures, cardData]}
            } else if (cardData.typeLine.toLowerCase().includes('land')) {
                return {...sortedByTypes, lands: [...sortedByTypes.lands, cardData]} 
            } else if (cardData.typeLine.toLowerCase().includes('enchantment')) {
                return {...sortedByTypes, enchantments: [...sortedByTypes.enchantments, cardData]}
            } else if (cardData.typeLine.toLowerCase().includes('artifact')) {
                return {...sortedByTypes, artifacts: [...sortedByTypes.artifacts, cardData]}
            } else if (cardData.typeLine.toLowerCase().includes('planeswalker')) {
                return {...sortedByTypes, planeswalkers: [...sortedByTypes.planeswalkers, cardData]}
            } else if (cardData.typeLine.toLowerCase().includes('sorcery')) {
                return {...sortedByTypes, spells: [...sortedByTypes.spells, cardData]}
            } else if (cardData.typeLine.toLowerCase().includes('instant')) {
                return {...sortedByTypes, spells: [...sortedByTypes.spells, cardData]}
            } else {
                return {...sortedByTypes, other: [...sortedByTypes.other, cardData]}
            }
        }, {
                artifacts: [],
                enchantments: [],
                spells: [],
                planeswalkers: [],
                lands: [],
                creatures: [],
                other: []
        })

        return (
            <div style={decklistStyle}>
                <div><b> DECK TITLE </b></div>
                <DeckTypeSelection />
                <table>
                    <DeckListEntries type={"Creatures"} data={sortedByTypes.creatures} counts={this.props.counts}/>
                    <DeckListEntries type={"Spells"} data={sortedByTypes.spells} counts={this.props.counts} />
                    <DeckListEntries type={"Enchantments"} data={sortedByTypes.enchantments} counts={this.props.counts}/>
                    <DeckListEntries type={"Artifacts"} data={sortedByTypes.artifacts} counts={this.props.counts}/>
                    <DeckListEntries type={"Planeswalkers"} data={sortedByTypes.planeswalkers} counts={this.props.counts}/>
                    <DeckListEntries type={"Lands"} data={sortedByTypes.lands} counts={this.props.counts}/>
                    <DeckListEntries type={"Other *debugging*"} data={sortedByTypes.other} counts={this.props.counts}/>
                </table>
            </div>
        )
    }
}

export default connect(mapStateToProps)(DeckList);