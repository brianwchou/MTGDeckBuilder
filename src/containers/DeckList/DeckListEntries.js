import React from 'react'
import DeckListEntry from 'containers/DeckList/DeckListEntry';

export default function DeckListEntries({data, type, counts, getCardInfo}) {
    let entries = data.map((entry, index)=> {
        return <DeckListEntry index={index} card={entry} getCardInfo={getCardInfo} count={counts[entry.name]} />
    }) 
    return (entries.length) ? (
        <>
            <div className="type_padding">{type}</div>
            <div>
                {entries}
            </div>
        </>
    ) : null;
}
