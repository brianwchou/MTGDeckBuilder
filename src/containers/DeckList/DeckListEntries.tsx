import React from 'react'
import DeckListEntry from './DeckListEntry'
import { CardInfo } from '../../common/types';

interface CardCount {
    [index: string]: number
}

type DeckListEntriesProps = {
    data: Array<CardInfo>,
    type: string,
    counts: CardCount,
    getCardInfo(card: CardInfo, name: string): void,
}

export default function DeckListEntries({data, type, counts, getCardInfo}: DeckListEntriesProps) {
    let entries = data.map((entry: CardInfo, index: number)=> {
        return <DeckListEntry key={index} card={entry} getCardInfo={getCardInfo} count={counts[entry.name]} />
    }) 
    return (entries.length) ? (
        <>
            <div className="type_padding">{type}</div>
            <div>
                {entries}
            </div>
        </>
    ) : null;
}
