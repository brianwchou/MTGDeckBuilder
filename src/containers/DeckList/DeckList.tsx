import React from 'react';
import { connect } from 'react-redux';
import Metrics from '../../components/Metrics';
import DeckTypeSelection from '../DeckList/DeckTypeSelection';
import DeckListEntries from '../DeckList/DeckListEntries';
import {incrementCardCount, decrementCardCount, moveToMaybe} from '../../actions/CardActions';
import './DeckList.css'
import { CardInfo, CardCount } from '../../common/types'
import { Dispatch } from 'react';
import { AnyAction } from 'redux';

export interface OrangizedCards {
  artifacts: CardInfo[],
  enchantments: CardInfo[],
  spells: CardInfo[],
  planeswalkers: CardInfo[],
  lands: CardInfo[],
  creatures: CardInfo[],
  other: CardInfo[]
}

export interface DeckListState {
  deckList: CardInfo[],
  cardCount: CardCount
}

export interface DeckListProps extends DeckListState {
  dispatch?: Dispatch<AnyAction>
}

export const mapStateToProps = (state: DeckListState) => {
  const {deckList, cardCount} = state
  return {
    deckList,
    cardCount
  }
}

export const organizeCards = (cards: CardInfo[]) => {
  return cards.reduce((sortedByTypes: OrangizedCards, cardData: CardInfo) => {
    const typeline: String = cardData.typeLine.toLowerCase()
    const {
      creatures, 
      lands, 
      enchantments, 
      artifacts, 
      planeswalkers, 
      spells, 
      other
    } = sortedByTypes

    if (typeline.includes('creature')) {
      return {...sortedByTypes, 
        creatures: [...creatures, cardData]}
    } else if (typeline.includes('land')) {
      return {...sortedByTypes, 
        lands: [...lands, cardData]} 
    } else if (typeline.includes('enchantment')) {
      return {...sortedByTypes, 
        enchantments: [...enchantments, cardData]}
    } else if (typeline.includes('artifact')) {
      return {...sortedByTypes, 
        artifacts: [...artifacts, cardData]}
    } else if (typeline.includes('planeswalker')) {
      return {...sortedByTypes, 
        planeswalkers: [...planeswalkers, cardData]}
    } else if (typeline.includes('sorcery')) {
      return {...sortedByTypes, 
        spells: [...spells, cardData]}
    } else if (typeline.includes('instant')) {
      return {...sortedByTypes, 
        spells: [...spells, cardData]}
    } else {
      return {...sortedByTypes, 
        other: [...other, cardData]}
    }
  }, {
    artifacts: [],
    enchantments: [],
    spells: [],
    planeswalkers: [],
    lands: [],
    creatures: [],
    other: []
  })
}

const decklistStyle = {
    border: 'solid black thin',
    width: '35vw',
    height: '50vh',
    padding: '15px',
    overflow: 'scroll',
}

export class DeckList extends React.Component<DeckListProps> {
  constructor(props) {
      super(props);
      this.getCardInfo = this.getCardInfo.bind(this);
  }

  getCardInfo(cardInfo, buttonType) {
    if (buttonType === 'increment') {
      this.props.dispatch(incrementCardCount(cardInfo));
    } else if (buttonType === 'decrement') {
      this.props.dispatch(decrementCardCount(cardInfo));
    } else if (buttonType === 'maybe') {
      this.props.dispatch(moveToMaybe(cardInfo));
    }
  }

  render() {
    // expecting deck lists sort data here?
    const sortedByTypes = organizeCards(this.props.deckList)
    const {
      creatures, 
      lands, 
      enchantments, 
      artifacts, 
      planeswalkers, 
      spells, 
      other
    } = sortedByTypes

    return (
      <div>
        <div style={decklistStyle}>
          <div><b> DECK TITLE </b></div>
          <DeckTypeSelection />
          <DeckListEntries getCardInfo={this.getCardInfo} type={"Creatures"} data={creatures} counts={this.props.cardCount}/>
          <DeckListEntries getCardInfo={this.getCardInfo} type={"Spells"} data={spells} counts={this.props.cardCount} />
          <DeckListEntries getCardInfo={this.getCardInfo} type={"Enchantments"} data={enchantments} counts={this.props.cardCount}/>
          <DeckListEntries getCardInfo={this.getCardInfo} type={"Artifacts"} data={artifacts} counts={this.props.cardCount}/>
          <DeckListEntries getCardInfo={this.getCardInfo} type={"Planeswalkers"} data={planeswalkers} counts={this.props.cardCount}/>
          <DeckListEntries getCardInfo={this.getCardInfo} type={"Lands"} data={lands} counts={this.props.cardCount}/>
          <DeckListEntries getCardInfo={this.getCardInfo} type={"Other *debugging*"} data={other} counts={this.props.cardCount}/>
        </div>
          
        <Metrics main={this.props.deckList} counts={this.props.cardCount}/>
      </div>
    )
  }
}

export default connect(mapStateToProps)(DeckList);