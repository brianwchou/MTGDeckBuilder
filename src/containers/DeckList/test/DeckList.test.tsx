import { DeckList, mapStateToProps, organizeCards } from '../../DeckList/DeckList'
import React from 'react'
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { DeckListState } from '../DeckList';
import { dummyCardStart, dummyCardUnchanged, dummyCardCount } from '../../../common/testUil';
import { CardInfo } from '../../../common/types';

configure({ adapter: new Adapter() });

describe('DeckList', () => {

    test('mapStateToProps', () => {
      const mockDeckList = [dummyCardStart]
      const mockCounts = dummyCardCount
      const mockState: DeckListState = { deckList: mockDeckList, cardCount: mockCounts }

      expect(mapStateToProps(mockState))
        .toEqual({ deckList: [ dummyCardUnchanged ], cardCount: dummyCardCount })
    });

    test('empty decklist renders correctly', () => {
        const mockDeckList: CardInfo[] = []
        const mockCounts = {}
        const wrapper = shallow(<DeckList deckList={mockDeckList} cardCount={mockCounts} />)

        expect(wrapper).toMatchSnapshot();
    });

    test('filled decklist renders correctly', () => {
        const mockMain = [dummyCardStart]
        const mockCounts = {}
        const wrapper = shallow(<DeckList deckList={mockMain} cardCount={mockCounts} />)

        expect(wrapper).toMatchSnapshot();
    });

    // TODO: refactor required
//     test('getCardInfo dispatches when button is clicked', () => {
//         const mockCounts = dummyCardCount
//         const mockMain = [dummyCardStart]
//         const onClickDispatch = jest.fn()
//         const wrapper = shallow<DeckList>(<DeckList deckList={mockMain} cardCount={mockCounts} dispatch={onClickDispatch} />)
//         const getCardInfoSpy = jest.spyOn(wrapper.instance(), 'getCardInfo')
//         // mount(<DeckList deckList={mockMain} cardCount={mockCounts} getCardInfo={onClickSpy} dispatch={onClickDispatch} />)

//         let container = wrapper.find("div")
//         let containerDeep = container.find("div")
//         let containerDeeper = containerDeep.find("div")
//         let containerButtonIncrement = containerDeeper.find("button").at(0)

//         expect(onClickDispatch).not.toHaveBeenCalled();
//         containerButtonIncrement.simulate('click');
//         expect(onClickDispatch).toHaveBeenCalled();

//         let containerButtonDecrement = containerDeeper.find("button").at(1)
        
//         onClickDispatch.mockClear();
//         expect(onClickDispatch).not.toHaveBeenCalled();
//         containerButtonDecrement.simulate('click');
//         expect(onClickDispatch).toHaveBeenCalled();

//         let containerButtonMaybe = containerDeeper.find("button").at(2)
        
//         onClickDispatch.mockClear();
//         expect(onClickDispatch).not.toHaveBeenCalled();
//         containerButtonMaybe.simulate('click');
//         expect(onClickDispatch).toHaveBeenCalled();
//     });
});

describe('organizeCards', () => {

  const mockCards = [
    { 
      artist: "",
      cmc: 2,
      color_identity: ["B"],
      colors: ["B"],
      image_uris: {
        small: "",
        normal: "",
        large: "",
        png: "",
        art_crop: "",
        border_crop: ""
      },
      mana_cost: "{1}{B}",
      name: "Cover of Darkness",
      oracle_text: "",
      power: undefined,
      rarity: "rare",
      reserved: false,
      setName: undefined,
      toughness: undefined,
      typeLine: ""
    }
  ]

  beforeEach(() => {
    mockCards[0].typeLine = ""
  })

  const testHelper = (result, typeline: string) => {
    Object.keys(result)
      .filter(key => key !== typeline)
      .forEach((key) => {
        expect(result[key]).toEqual([])
      })

    expect(result[typeline]).toHaveLength(1)
  }

  test('sorting creature card', () => {
    const testedTypeline = "creatures"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)
    
    testHelper(result, testedTypeline)
  });

  test('sorting land card', () => {
    const testedTypeline = "lands"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)

    testHelper(result, testedTypeline)
  });

  test('sorting enchantment card', () => {
    const testedTypeline = "enchantments"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)
    
    testHelper(result, testedTypeline)
  });

  test('sorting artifact card', () => {
    const testedTypeline = "artifacts"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)
    
    testHelper(result, testedTypeline)
  });

  test('sorting planeswalker card', () => {
    const testedTypeline = "planeswalkers"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)
    
    testHelper(result, testedTypeline)
  });

  test('sorting sorcery card', () => {
    const testedTypeline = "sorcery"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)
    
    testHelper(result, "spells")
  });

  test('sorting instant card', () => {
    const testedTypeline = "instant"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)
    
    testHelper(result, "spells")
  });

  test('sorting card with none of the specified categories', () => {
    const testedTypeline = "hgdfjdshgjfsh"
    mockCards[0].typeLine = testedTypeline

    let result = organizeCards(mockCards)
    
    testHelper(result, "other")
  });
});