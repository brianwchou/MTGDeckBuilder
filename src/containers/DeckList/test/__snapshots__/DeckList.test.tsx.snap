// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckList empty decklist renders correctly 1`] = `ShallowWrapper {}`;

exports[`DeckList filled decklist renders correctly 1`] = `ShallowWrapper {}`;
